lib/api/apiUtils/bucket/bucketEncryption.js

```diff
@@ -83,32 +94,128 @@ function parseEncryptionXml(xml, log, cb) {
+        sseConfig.mandatory = mandatory;
+    }
+
+    return sseConfig;
+}
+
 /**
  * parseBucketEncryptionHeaders - retrieves bucket level sse configuration from request headers
  * @param {object} headers - Request headers
- * @returns {object} SSE configuration
+ * @returns {serverSideEncryptionInfo} - SSE configuration
  */
 function parseBucketEncryptionHeaders(headers) {
     const sseAlgorithm = headers['x-amz-scal-server-side-encryption'];
     const configuredMasterKeyId = headers['x-amz-scal-server-side-encryption-aws-kms-key-id'] || null;
+    return hydrateEncryptionConfig(sseAlgorithm, configuredMasterKeyId, true);
+}
 
-    if (sseAlgorithm === 'AES256') {
-        return {
-            algorithm: sseAlgorithm,
-        };
-    } else if (sseAlgorithm === 'aws:kms') {
-        return {
-            algorithm: sseAlgorithm,
-            configuredMasterKeyId,
-        };
+/**
+ * parseObjectEncryptionHeaders - retrieves bucket level sse configuration from request headers
+ * @param {object} headers - Request headers
+ * @returns {serverSideEncryptionInfo} - SSE configuration
+ */
+function parseObjectEncryptionHeaders(headers) {
+    const sseAlgorithm = headers['x-amz-server-side-encryption'];
+    const configuredMasterKeyId = headers['x-amz-server-side-encryption-aws-kms-key-id'] || null;
+    return hydrateEncryptionConfig(sseAlgorithm, configuredMasterKeyId);
+}
+
+/**
+ * createDefaultBucketEncryptionMetadata - Creates master key and sets up default server side encryption configuration
+ * @param {BucketInfo} bucket - bucket metadata
+ * @param {object} log - werelogs logger
+ * @param {*} cb - callback
+ * @returns {undefined}
+ */
+function createDefaultBucketEncryptionMetadata(bucket, log, cb) {
+    return kms.bucketLevelEncryption(
+        bucket.getName(),
+        { algorithm: 'AES256', mandatory: false },
+        log,
+        (error, sseConfig) => {
+            if (error) {
+                return cb(error);
+            }
+            bucket.setServerSideEncryption(sseConfig);
+            return metadata.updateBucket(bucket.getName(), bucket, log, err => cb(err, sseConfig));
+        });
+}
+
+/**
+ *
+ * @param {object} headers - request headers
+ * @param {BucketInfo} bucket - BucketInfo model
+ * @param {*} log - werelogs logger
+ * @param {*} cb - callback
+ * @returns {undefined}
+ */
+function getObjectSSEConfiguration(headers, bucket, log, cb) {
+    const objectSSE = parseObjectEncryptionHeaders(headers);
+    const bucketSSE = bucket.getServerSideEncryption();
+
+    // If a per object sse algo has been passed through
+    // x-amz-server-side-encryption
+    if (objectSSE.algorithm) {
+        // If aws:kms and a custom key id
+        // pass it through without updating the bucket md
+        if (objectSSE.algorithm === 'aws:kms' && objectSSE.configuredMasterKeyId) {
+            return cb(null, objectSSE);
+        }
+
+        // If the client has not specified a key id,
```

Comment:
I believed that a client passing `aws:kms` as algorithm without an `aws-kms-key-id` would be considered equivalent to `AES256` algorithm, but I may be wrong. Here we seem to consider it as invalid and we ignore the header, it could be correct just wondering, I don't remember exactly what the spec says.

Reply:
When a request specifies `aws:kms` it will use the buckets existing `masterKeyId` to encrypt if it exists (thats what its doing here). We do probably need to change the algorithm value to `aws:kms` so an get-object's headers are correct